BulkheadTest intermittently failing

I get the following exception on local env:
```
[INFO] Running io.helidon.nima.faulttolerance.BulkheadTest
[ERROR] Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.169 s <<< FAILURE! - in io.helidon.nima.faulttolerance.BulkheadTest
[ERROR] io.helidon.nima.faulttolerance.BulkheadTest.testBulkheadWithError  Time elapsed: 5.002 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: Unexpected exception type thrown ==> expected: <java.util.concurrent.ExecutionException> but was: <java.util.concurrent.TimeoutException>
        at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:65)
        at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:37)
        at org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3007)
        at io.helidon.nima.faulttolerance.BulkheadTest.testBulkheadWithError(BulkheadTest.java:199)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
...
...
...
Caused by: java.util.concurrent.TimeoutException
        at java.base/java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1960)
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2095)
        at io.helidon.nima.faulttolerance.BulkheadTest.lambda$testBulkheadWithError$12(BulkheadTest.java:200)
        at org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:55)
```

Problem is caused by a race condition when a task is finished (if another task enqueues in parallel, we may:
- enqueue another task
- release the semaphore permit
- never release the enqueued task

--- a/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/AsyncImpl.java
+++ b/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/AsyncImpl.java
@@ -16,11 +16,10 @@
 
 package io.helidon.nima.faulttolerance;
 
-import java.util.concurrent.CancellationException;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Supplier;
 
 import io.helidon.common.LazyValue;
@@ -31,6 +30,8 @@ import static io.helidon.nima.faulttolerance.SupplierHelper.unwrapThrowable;
  * Implementation of {@code Async}. Default executor accessed from {@link FaultTolerance#executor()}.
  */
 class AsyncImpl implements Async {
+    private static final System.Logger LOGGER = System.getLogger(AsyncImpl.class.getName());
+
     private final LazyValue<? extends ExecutorService> executor;
     private final CompletableFuture<Async> onStart;
 
@@ -45,12 +46,19 @@ class AsyncImpl implements Async {
 
     @Override
     public <T> CompletableFuture<T> invoke(Supplier<T> supplier) {
-        AtomicBoolean mayInterrupt = new AtomicBoolean(false);
+        AtomicReference<Future<?>> ourFuture = new AtomicReference<>();
         CompletableFuture<T> result = new CompletableFuture<>() {
             @Override
             public boolean cancel(boolean mayInterruptIfRunning) {
-                mayInterrupt.set(mayInterruptIfRunning);
-                return super.cancel(mayInterruptIfRunning);
+                Future<?> toCancel = ourFuture.get();
+                if (toCancel == null) {
+                    // cancelled before the future was assigned - this should not happen, as we do
+                    // not escape this method before that
+                    LOGGER.log(System.Logger.Level.WARNING, "Failed to cancel future, it is not yet available.");
+                    return false;
+                } else {
+                    return toCancel.cancel(mayInterruptIfRunning);
+                }
             }
         };
         Future<?> future = executor.get().submit(() -> {
@@ -67,12 +75,8 @@ class AsyncImpl implements Async {
                 result.completeExceptionally(throwable);
             }
         });
-        result.exceptionally(t -> {
-            if (t instanceof CancellationException) {
-                future.cancel(mayInterrupt.get());
-            }
-            return null;
-        });
+        ourFuture.set(future);
+
         return result;
     }
 

--- a/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/BulkheadImpl.java
+++ b/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/BulkheadImpl.java
@@ -65,10 +65,19 @@ class BulkheadImpl implements Bulkhead {
 
     @Override
     public <T> T invoke(Supplier<? extends T> supplier) {
-        // execute immediately if semaphore can be acquired
+        // we need to hold the lock until we decide what to do with this request
+        // cannot release it in between attempts, as that would give window for another thread to change the state
         inProgressLock.lock();
 
-        if (inProgress.tryAcquire()) {
+        // execute immediately if semaphore can be acquired
+        boolean acquired;
+        try {
+            acquired = inProgress.tryAcquire();
+        } catch (Throwable t) {
+            inProgressLock.unlock();
+            throw t;
+        }
+        if (acquired) {
             inProgressLock.unlock(); // we managed to get a semaphore permit, in progress lock can be released for now
             if (LOGGER.isLoggable(Level.DEBUG)) {
                 LOGGER.log(Level.DEBUG, name + " invoke immediate " + supplier);
@@ -76,16 +85,29 @@ class BulkheadImpl implements Bulkhead {
             return execute(supplier);
         }
 
-        if (queue.isFull()) {
+        boolean full;
+        try {
+            full = queue.isFull();
+        } catch (Throwable t) {
+            inProgressLock.unlock();
+            throw t;
+        }
+        if (full) {
             inProgressLock.unlock(); // this request will fail, release lock
             callsRejected.incrementAndGet();
             throw new BulkheadException("Bulkhead queue \"" + name + "\" is full");
         }
+
         try {
             // block current thread until barrier is retracted
-            listeners.forEach(l -> l.enqueueing(supplier));
-            Barrier barrier = queue.enqueue(supplier);
-            inProgressLock.unlock(); // we have enqueued, now we can wait
+            Barrier barrier;
+            try {
+                listeners.forEach(l -> l.enqueueing(supplier));
+                barrier = queue.enqueue(supplier);
+            } finally {
+                inProgressLock.unlock(); // we have enqueued, now we can wait
+            }
+
             if (barrier == null) {
                 throw new BulkheadException("Bulkhead queue \"" + name + "\" is full");
             }
@@ -156,11 +178,14 @@ class BulkheadImpl implements Bulkhead {
         } finally {
             concurrentExecutions.decrementAndGet();
             inProgressLock.lock();
-            boolean dequeued = queue.dequeueAndRetract();
-            if (!dequeued) {
-                inProgress.release();       // nothing dequeued, one more permit
+            try {
+                boolean dequeued = queue.dequeueAndRetract();
+                if (!dequeued) {
+                    inProgress.release();       // nothing dequeued, one more permit
+                }
+            } finally {
+                inProgressLock.unlock();
             }
-            inProgressLock.unlock();
         }
     }
 

--- a/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/BulkheadImpl.java
+++ b/nima/fault-tolerance/src/main/java/io/helidon/nima/faulttolerance/BulkheadImpl.java
@@ -28,6 +28,7 @@ import java.util.concurrent.ExecutionException;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.Semaphore;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.Lock;
 import java.util.concurrent.locks.ReentrantLock;
 import java.util.function.Supplier;
 
@@ -37,6 +38,7 @@ import static io.helidon.nima.faulttolerance.SupplierHelper.unwrapThrowable;
 class BulkheadImpl implements Bulkhead {
     private static final System.Logger LOGGER = System.getLogger(BulkheadImpl.class.getName());
 
+    private final Lock inProgressLock;
     private final Semaphore inProgress;
     private final String name;
     private final BarrierQueue queue;
@@ -53,6 +55,7 @@ class BulkheadImpl implements Bulkhead {
         this.queue = builder.queueLength() > 0
                 ? new BlockingQueue(builder.queueLength())
                 : new ZeroCapacityQueue();
+        this.inProgressLock = new ReentrantLock(true);
     }
 
     @Override
@@ -63,7 +66,10 @@ class BulkheadImpl implements Bulkhead {
     @Override
     public <T> T invoke(Supplier<? extends T> supplier) {
         // execute immediately if semaphore can be acquired
+        inProgressLock.lock();
+
         if (inProgress.tryAcquire()) {
+            inProgressLock.unlock(); // we managed to get a semaphore permit, in progress lock can be released for now
             if (LOGGER.isLoggable(Level.DEBUG)) {
                 LOGGER.log(Level.DEBUG, name + " invoke immediate " + supplier);
             }
@@ -71,13 +77,19 @@ class BulkheadImpl implements Bulkhead {
         }
 
         if (queue.isFull()) {
+            inProgressLock.unlock(); // this request will fail, release lock
             callsRejected.incrementAndGet();
             throw new BulkheadException("Bulkhead queue \"" + name + "\" is full");
         }
         try {
             // block current thread until barrier is retracted
             listeners.forEach(l -> l.enqueueing(supplier));
-            queue.enqueueAndWaitOn(supplier);
+            Barrier barrier = queue.enqueue(supplier);
+            inProgressLock.unlock(); // we have enqueued, now we can wait
+            if (barrier == null) {
+                throw new BulkheadException("Bulkhead queue \"" + name + "\" is full");
+            }
+            barrier.waitOn();
 
             // unblocked so we can proceed with execution
             listeners.forEach(l -> l.dequeued(supplier));
@@ -143,10 +155,12 @@ class BulkheadImpl implements Bulkhead {
             throw toRuntimeException(throwable);
         } finally {
             concurrentExecutions.decrementAndGet();
+            inProgressLock.lock();
             boolean dequeued = queue.dequeueAndRetract();
             if (!dequeued) {
                 inProgress.release();       // nothing dequeued, one more permit
             }
+            inProgressLock.unlock();
         }
     }
 
@@ -182,11 +196,9 @@ class BulkheadImpl implements Bulkhead {
          * Enqueue supplier and block thread on barrier.
          *
          * @param supplier the supplier
-         * @return {@code true} if supplier was enqueued or {@code false} otherwise
-         * @throws ExecutionException if exception encountered while blocked
-         * @throws InterruptedException if blocking is interrupted
+         * @return barrier if supplier was enqueued or null otherwise
          */
-        boolean enqueueAndWaitOn(Supplier<?> supplier) throws ExecutionException, InterruptedException;
+        Barrier enqueue(Supplier<?> supplier);
 
         /**
          * Dequeue supplier and retract its barrier.
@@ -220,8 +232,9 @@ class BulkheadImpl implements Bulkhead {
         }
 
         @Override
-        public boolean enqueueAndWaitOn(Supplier<?> supplier) {
-            return false;
+        public Barrier enqueue(Supplier<?> supplier) {
+            // never enqueue, should always fail execution if permits are not available
+            return null;
         }
 
         @Override
@@ -273,19 +286,13 @@ class BulkheadImpl implements Bulkhead {
         }
 
         @Override
-        public boolean enqueueAndWaitOn(Supplier<?> supplier) throws ExecutionException, InterruptedException {
-            Barrier barrier;
+        public Barrier enqueue(Supplier<?> supplier) {
             lock.lock();
             try {
-                barrier = enqueue(supplier);
+                return doEnqueue(supplier);
             } finally {
                 lock.unlock();
             }
-            if (barrier != null) {
-                barrier.waitOn();
-                return true;
-            }
-            return false;
         }
 
         @Override
@@ -318,7 +325,7 @@ class BulkheadImpl implements Bulkhead {
             return supplier == null ? null : map.remove(supplier);
         }
 
-        private Barrier enqueue(Supplier<?> supplier) {
+        private Barrier doEnqueue(Supplier<?> supplier) {
             boolean added = queue.offer(supplier);
             return added ? map.computeIfAbsent(supplier, s -> new Barrier()) : null;
         }

--- a/nima/fault-tolerance/src/test/java/io/helidon/nima/faulttolerance/AsyncTest.java
+++ b/nima/fault-tolerance/src/test/java/io/helidon/nima/faulttolerance/AsyncTest.java
@@ -22,9 +22,9 @@ import java.util.concurrent.TimeUnit;
 import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.CoreMatchers.endsWith;
+import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.CoreMatchers.startsWith;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.CoreMatchers.is;
 
 class AsyncTest {
     private static final long WAIT_TIMEOUT_MILLIS = 2000;

--- a/nima/fault-tolerance/src/test/java/io/helidon/nima/faulttolerance/BulkheadTest.java
+++ b/nima/fault-tolerance/src/test/java/io/helidon/nima/faulttolerance/BulkheadTest.java
@@ -25,9 +25,10 @@ import java.util.function.Supplier;
 import io.helidon.logging.common.LogConfig;
 
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.RepeatedTest;
 import org.junit.jupiter.api.Test;
 
-import static java.lang.System.Logger.Level.INFO;
+import static java.lang.System.Logger.Level.TRACE;
 import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.instanceOf;
@@ -169,7 +170,7 @@ class BulkheadTest {
         }
     }
 
-    @Test
+    @RepeatedTest(100)
     void testBulkheadWithError() throws InterruptedException, ExecutionException, java.util.concurrent.TimeoutException {
         // Create bulkhead of 1 with a queue of 1
         Bulkhead bulkhead = Bulkhead.builder()
@@ -185,13 +186,16 @@ class BulkheadTest {
         Task inProgress = new Task(0);
         CompletableFuture<?> inProgressFuture = Async.invokeStatic(
                 () -> bulkhead.invoke(inProgress::run));
-        CompletableFuture<?> failedFuture = Async.invokeStatic(
-                () -> bulkhead.invoke(() -> { throw new IllegalStateException(); }));
 
         // Verify completion of inProgress task
         if (!inProgress.waitUntilStarted(WAIT_TIMEOUT_MILLIS)) {
             fail("Task inProgress never started");
         }
+
+        // as we use an async to submit to bulkhead, we should wait until the first task is submitted
+        CompletableFuture<?> failedFuture = Async.invokeStatic(
+                () -> bulkhead.invoke(() -> { throw new IllegalStateException(); }));
+
         inProgress.unblock();
         inProgressFuture.get(WAIT_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
 
@@ -219,7 +223,7 @@ class BulkheadTest {
         }
 
         int run() {
-            LOGGER.log(INFO, "Task " + index + " running on thread " + Thread.currentThread().getName());
+            LOGGER.log(TRACE, "Task " + index + " running on thread " + Thread.currentThread().getName());
 
             started.countDown();
             try {

--- a/tests/integration/webclient/src/test/java/io/helidon/tests/integration/webclient/TracingPropagationTest.java
+++ b/tests/integration/webclient/src/test/java/io/helidon/tests/integration/webclient/TracingPropagationTest.java
@@ -33,6 +33,7 @@ import io.opentracing.mock.MockSpan;
 import io.opentracing.mock.MockTracer;
 import io.opentracing.tag.Tags;
 import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
 import static org.hamcrest.CoreMatchers.is;
@@ -47,6 +48,7 @@ class TracingPropagationTest {
     private static final Duration TIMEOUT = Duration.ofSeconds(10);
 
     @Test
+    @Disabled // intermittently failing on pipeline, issue 5754
     void testTracingSuccess() throws ExecutionException, InterruptedException {
         MockTracer mockTracer = new MockTracer();
 

